stlab/channel.hpp

```diff
@@ -345,28 +386,49 @@ struct shared_process : shared_process_receiver<yield_type<T, Arg>>,
             } else {
                 message = std::move(_process_message_queue.front());
                 _process_message_queue.pop_front();
-                cts = _process_message_queue.size() == 0; // TODO (sparent) : queue size - 1.
+                do_cts = _process_message_queue.size() <= (_process_buffer_size - 1);
             }
         }
-        if (cts && _upstream) _upstream->cts();
+
+        if (do_cts && !_upstream.empty()) {
+            for (auto& up : _upstream) { up->clear_to_send(); }
+        }
+
         if (message) _process.await(std::move(message.get()));
         else if (do_close) process_close(_process);
         return bool(message);
     }
 
     template <typename U>
     auto step() -> std::enable_if_t<has_process_yield<U>> {
-        /*
-            REVISIT (sparent) : Put a timer on this loop to limit it?
-        */
-        while (get_process_state(_process) != process_state::yield) {
-            if (!dequeue()) break;
+        if (has_process_await_timeout<U> && get_process_await_timeout(_process) > std::chrono::milliseconds(0)) {
+            auto time_over = false;
+
+            auto ellapsing_point = std::chrono::system_clock::now() + get_process_await_timeout(_process);
+
+            while (!time_over && (get_process_state(_process) != process_state::yield)) {
```

Comment:
This effectively turns into an expensive spin-loop. If we didn't timeout and the queue is empty we will return and end up getting called again immediately.

If we are in an await state with a timeout, and the queue is empty, then schedule a task with the timeout (this requires a change to the scheduler interface) which when it fires will check to make sure it hasn't been canceled (by sent value triggering a call to await, or by a call to close) and will then invoke yield.